### ISSUE TITLE
(Suspected) regression in bytecode dump handling routines

### ISSUE THREAD

**Issue by pfalcon** [micropython/micropython#4341]
I'd like again to attention to particular commits of https://github.com/micropython/micropython/pull/4265 , specifically "tools/mpy-tools.py: Fix opcode size enumeration and constant definition." and "py/bc.c: Fix opcode size enumeration.", current links:

* https://github.com/micropython/micropython/pull/4265/commits/94fffddb19c2ab6a196c72f23aed85bc79d758df
* https://github.com/micropython/micropython/pull/4265/commits/d24402f4f1199cafa72d40b33f459dd397a9a4a9

And comment https://github.com/micropython/micropython/pull/4265#issuecomment-435183103.

What those commits show is that, allegedly, routines for decoding bytecode (as used for example by -v -v flags) have a bug/weren't updated after MICROPY_OPT_CACHE_MAP_LOOKUP_IN_BYTECODE was introduced, and tools/mpy-tool.py even has incorrect values for some bytecodes.

This is pretty serious regression, which, if true, may seriously confuse anyone trying to use MicroPython in advanced way, e.g. to look into optimizations.

Note that I didn't try to verify those findings (apparently, it's maintainer's job), hence modals sprinkled around the text.

**Comment by pfalcon** [comment]
> According to emitbc.c and bc.c MP_BC_LOAD_NAME and related opcodes can
have one extra byte when MICROPY_OPT_CACHE_MAP_LOOKUP_IN_BYTECODE is
enabled.

~~~
+    int extra_byte = (
+        *ip == MP_BC_RAISE_VARARGS
+        || *ip == MP_BC_MAKE_CLOSURE
+        || *ip == MP_BC_MAKE_CLOSURE_DEFARGS
+        #if MICROPY_OPT_CACHE_MAP_LOOKUP_IN_BYTECODE
+        || *ip == MP_BC_LOAD_NAME
+        || *ip == MP_BC_LOAD_GLOBAL
+        || *ip == MP_BC_LOAD_ATTR
+        || *ip == MP_BC_STORE_ATTR
+        #endif
+    );
     if (f == MP_OPCODE_QSTR) {
-        ip += 3;
+        ip += 3 + extra_byte;
     } else {
-        int extra_byte = (
-            *ip == MP_BC_RAISE_VARARGS
-            || *ip == MP_BC_MAKE_CLOSURE
-            || *ip == MP_BC_MAKE_CLOSURE_DEFARGS
-            #if MICROPY_OPT_CACHE_MAP_LOOKUP_IN_BYTECODE
-            || *ip == MP_BC_LOAD_NAME
-            || *ip == MP_BC_LOAD_GLOBAL
-            || *ip == MP_BC_LOAD_ATTR
-            || *ip == MP_BC_STORE_ATTR
-            #endif
-        );
~~~

**Comment by dpgeorge** [comment]
Thanks for bringing this up.  But it's not technically a regression because caching in the bytecode was added first, then frozen bytecode, then support for frozen bytecode with caching.

Just for the record, to understand what could be affected, the timeline is:
- Jan 2015: bytecode caching 7ee91cf86136f4774a9143a3f745d37f314662da
- Jan 2016: frozen .mpy files (not supporting caching) 0699c6bf9e7a1754002948063390b68590fc3e96
- Jan 2017: support for caching in frozen bytecode 98458a46ec9c34fa59dbfa7c9189b7e181e5e7a3

It's unlikely that any embedded targets use bytecode caching and frozen bytecode because in such a case the frozen bytecode must be stored in RAM, which really limits its usefulness.

**Comment by pfalcon** [comment]
> But it's not technically a regression

I stand corrected, and I used all those modals here because I indeed didn't try to investigate it myself (and don't remember all those details by now). All things bytecode are otherwise rather underused in MicroPython, some people show interest in knowing more about it, and having such issues could puzzle them a lot on their learning curve. Thanks for looking into this!
